### PR TITLE
IDE-3038 make new jsf Wizard for DXP

### PR DESCRIPTION
**Pull request by simonjhy**

### CODE REVIEW

File: tools/plugins/com.liferay.ide.scripting.core/.project [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.scripting.core/.project]: This should not be added in this pull.

File: tools/plugins/com.liferay.ide.project.ui/src/com/liferay/ide/project/ui/modules/NewLiferayJSFModuleProjectWizard.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    implements IWorkbenchWizard, INewWizard
+{
+    private boolean firstErrorMessageRemoved = false;
+
+    public NewLiferayJSFModuleProjectWizard()
+    {
+        super( createDefaultOp(), DefinitionLoader.sdef( NewLiferayJSFModuleProjectWizard.class ).wizard() );
+    }
+
+    private void addToWorkingSets( IProject newProject ) throws Exception
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.project.ui/src/com/liferay/ide/project/ui/modules/NewLiferayJSFModuleProjectWizard.java]: this should be shared with other wizards

File: tools/plugins/com.liferay.ide.project.ui/src/com/liferay/ide/project/ui/modules/NewLiferayJSFModuleProjectWizard.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                    {
+                        PlatformUI.getWorkbench().getWorkingSetManager().addToWorkingSets(newProject, workingSets);
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public IWizardPage[] getPages()
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.project.ui/src/com/liferay/ide/project/ui/modules/NewLiferayJSFModuleProjectWizard.java]: This should be shared with other wizards

File: tools/plugins/com.liferay.ide.project.ui/src/com/liferay/ide/project/ui/modules/NewLiferayJSFModuleProjectWizard.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+                if( location != null )
+                {
+                    element().setInitialSelectionPath( PathBridge.create( location ) );
+                }
+            }
+        }
+    }
+
+    private void openLiferayPerspective( IProject newProject )
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.project.ui/src/com/liferay/ide/project/ui/modules/NewLiferayJSFModuleProjectWizard.java]: this should be shared with other wizards

File: tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/util/LiferayWorkspaceUtil.java [hunk 5]
```diff
@@ -356,4 +361,114 @@ public static IPath getHomeLocation( IProject project )
...
+        {
+            Properties properties = PropertiesUtil.loadProperties( gradleProperties );
+
+            retVal = properties.getProperty( key, defaultValue );
+        }
+
+        return retVal;
+    }
+
+    public static boolean isWorkspace( File dir )
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/util/LiferayWorkspaceUtil.java]: we already have other methods that check if it is workspace.  Why did you need to write another one?

File: tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/fragment/FragmentProjectNameValidationService.java [hunk 6]
```diff
@@ -15,135 +15,51 @@
...
-            this.listener = null;
-        }
-
-        super.dispose();
+        return NewModuleFragmentOp.PROP_PROJECT_NAME;
     }
 
-    private boolean isValidProjectName( String currentProjectName )
+    @Override
+    protected <A> A getData( String key, Class<A> type )
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/fragment/FragmentProjectNameValidationService.java]: This looks incorrect.  We shouldn't override getData and then have to hard-code strings for "projectName" or "Location"  that will break if we ever refactored those properties

File: tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/fragment/FragmentProjectLocationValidationService.java [hunk 7]
```diff
@@ -15,141 +15,54 @@
...
+    protected <A> A getData( String key, Class<A> type )
     {
-        if( this.listener != null && op() != null && !op().disposed() )
-        {
-            op().getProjectName().detach( this.listener );
-            op().getProjectProvider().detach( this.listener );
+        A retval = null;
 
-            this.listener = null;
+        if( "ProjectName".equals( key ) )
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/fragment/FragmentProjectLocationValidationService.java]: This code is not good.

File: tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/ModuleProjectNameValidationService.java [hunk 8]
```diff
@@ -14,127 +14,47 @@
...
-
     @Override
-    protected void initValidationService()
+    protected <A> A getData( String key, Class<A> type )
     {
-        super.initValidationService();
+        A retval = null;
 
-        listener = new FilteredListener<PropertyContentEvent>()
+        if( "ProjectName".equals( key ) )
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/ModuleProjectNameValidationService.java]: Lets not do this

File: tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/AbstractProjectLocationValidationService.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+        return file.canWrite();
+    }
+
+    @Override
+    protected Status compute()
+    {
+        Status retval = Status.createOkStatus();
+
+        final Path currentProjectLocation = getData( "Location", Path.class );
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/AbstractProjectLocationValidationService.java]: We should be using real properties not looking them up by string names.  that will break if we change variable names

File: tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/jsf/JSFModuleTemplatePossibleValuesService.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+public class JSFModuleTemplatePossibleValuesService extends PossibleValuesService
+{
+
+    private List<String> possibleValues;
+
+    protected void initPossibleValuesService()
+    {
+        possibleValues = new ArrayList<String>();
+
+        for( String key : NewLiferayJSFModuleProjectOpMethods.JSF_TEMPLATE_MAP.keySet() )
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/jsf/JSFModuleTemplatePossibleValuesService.java]: The possible values should not be the Display names.  The possible values should be the lowercase names "jsf" "primefaces" and then we should have a "LabelPresentationProvider" that converts from "primefaces" to "PrimeFaces" That way we don't need the `JSF_TEMPLATE_MAP` at all.  Look through our other code for "Label" Provider

File: tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/jsf/JSFModuleProjectNameValidationService.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        return context( NewLiferayJSFModuleProjectOp.class );
+    }
+
+    @Override
+    protected <A> A getData( String key, Class<A> type )
+    {
+        A retval = null;
+
+        if( "ProjectName".equals( key ) )
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/jsf/JSFModuleProjectNameValidationService.java]: Don't do this type of code.

File: tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/jsf/JSFModuleProjectNameListener.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+                if( workspaceDir == null || !workspaceDir.exists() )
+                {
+                    return;
+                }
+                String[] folders =
+                    LiferayWorkspaceUtil.getLiferayWorkspaceProjectWarsDirs( workspaceDir.getAbsolutePath() );
+
+                if( folders != null )
+                {
+                    boolean appedndWarFolder = false;
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/jsf/JSFModuleProjectNameListener.java]: should be 'appendedWarFolder'

File: tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/jsf/JSFModuleProjectLocationValidationService.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+        op().getProjectName().attach( listener );
+        op().getProjectProvider().attach( listener );
+    }
+
+    @Override
+    protected <A> A getData( String key, Class<A> type )
+    {
+        A retval = null;
+
+        if( "ProjectName".equals( key ) )
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/jsf/JSFModuleProjectLocationValidationService.java]: Don't do this type of code

File: maven/plugins/com.liferay.ide.maven.core/src/com/liferay/ide/maven/core/NewMavenJSFModuleProjectProvider.java [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+            {
+                return LiferayCore.createErrorStatus( "The project location of Liferay Workspace shoule be existed." );
+            }
+
+            String[] folders =
+                LiferayWorkspaceUtil.getLiferayWorkspaceProjectWarsDirs( workspaceDir.getAbsolutePath() );
+
+            if( folders != null )
+            {
+                boolean appedndWarFolder = false;
```
Comment (gamerson) [inline maven/plugins/com.liferay.ide.maven.core/src/com/liferay/ide/maven/core/NewMavenJSFModuleProjectProvider.java]: `appendedWarFolder`

File: maven/plugins/com.liferay.ide.maven.core/src/com/liferay/ide/maven/core/NewGradleJSFModuleProjectProvider.java [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+import org.eclipse.wst.validation.internal.provisional.core.IValidatorJob;
+
+/**
+ * @author Simon Jiang
+ */
+
+@SuppressWarnings( "restriction" )
+public class NewGradleJSFModuleProjectProvider extends NewMavenJSFModuleProjectProvider
+{
+    public final static String VALIDATOR_ID = "org.eclipse.wst.html.core.HTMLValidator";
```
Comment (gamerson) [inline maven/plugins/com.liferay.ide.maven.core/src/com/liferay/ide/maven/core/NewGradleJSFModuleProjectProvider.java]: What is this for?  It is a variable that isn't used?

File: tools/plugins/com.liferay.ide.project.ui/src/com/liferay/ide/project/ui/modules/NewLiferayJSFModuleProjectWizard.java (hunk 2 above)
Comment (simonjhy) [inline tools/plugins/com.liferay.ide.project.ui/src/com/liferay/ide/project/ui/modules/NewLiferayJSFModuleProjectWizard.java]: Done

File: tools/plugins/com.liferay.ide.project.ui/src/com/liferay/ide/project/ui/modules/NewLiferayJSFModuleProjectWizard.java (hunk 3 above)
Comment (simonjhy) [inline tools/plugins/com.liferay.ide.project.ui/src/com/liferay/ide/project/ui/modules/NewLiferayJSFModuleProjectWizard.java]: Done

File: tools/plugins/com.liferay.ide.project.ui/src/com/liferay/ide/project/ui/modules/NewLiferayJSFModuleProjectWizard.java (hunk 4 above)
Comment (simonjhy) [inline tools/plugins/com.liferay.ide.project.ui/src/com/liferay/ide/project/ui/modules/NewLiferayJSFModuleProjectWizard.java]: Done

File: tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/util/LiferayWorkspaceUtil.java (hunk 5 above)
Comment (simonjhy) [inline tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/util/LiferayWorkspaceUtil.java]: ok, I  reused other method.

File: tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/fragment/FragmentProjectNameValidationService.java (hunk 6 above)
Comment (simonjhy) [inline tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/fragment/FragmentProjectNameValidationService.java]: ok, changed to get property name from op.

File: tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/fragment/FragmentProjectLocationValidationService.java (hunk 7 above)
Comment (simonjhy) [inline tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/fragment/FragmentProjectLocationValidationService.java]: changed to get property name from op.

File: tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/ModuleProjectNameValidationService.java (hunk 8 above)
Comment (simonjhy) [inline tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/ModuleProjectNameValidationService.java]: changed to get property name from op.

File: maven/plugins/com.liferay.ide.maven.core/src/com/liferay/ide/maven/core/NewGradleJSFModuleProjectProvider.java (hunk 15 above)
Comment (simonjhy) [inline maven/plugins/com.liferay.ide.maven.core/src/com/liferay/ide/maven/core/NewGradleJSFModuleProjectProvider.java]: I tried to use it disable htmlValidator check and want to resolve gradle Jsf project validation problem I asked you yesterday. I will remove it.
